internal/querynodev2/pkoracle/pk_oracle.go

```diff
@@ -41,18 +43,46 @@ var _ PkOracle = (*pkOracle)(nil)
 // pkOracle implementation.
 type pkOracle struct {
 	candidates *typeutil.ConcurrentMap[string, candidateWithWorker]
+
+	hashFuncNumMutex sync.RWMutex
+	maxHashFuncNum   uint
+}
+
+func (pko *pkOracle) GetMaxHashFuncNum() uint {
+	pko.hashFuncNumMutex.RLock()
+	defer pko.hashFuncNumMutex.RUnlock()
+	return pko.maxHashFuncNum
+}
+
+func (pko *pkOracle) TryUpdateHashFuncNum(newValue uint) {
+	pko.hashFuncNumMutex.Lock()
+	defer pko.hashFuncNumMutex.Unlock()
+	if newValue > pko.maxHashFuncNum {
+		pko.maxHashFuncNum = newValue
+	}
 }
 
 // Get implements PkOracle.
 func (pko *pkOracle) Get(pk storage.PrimaryKey, filters ...CandidateFilter) ([]int64, error) {
 	var result []int64
+	var locations []uint64
+
 	pko.candidates.Range(func(key string, candidate candidateWithWorker) bool {
 		for _, filter := range filters {
 			if !filter(candidate) {
 				return true
 			}
 		}
-		if candidate.MayPkExist(pk) {
+
+		if locations == nil {
+			locations = storage.Locations(pk, pko.GetMaxHashFuncNum())
+			if len(locations) == 0 {
+				log.Warn("pkOracle: no location found for pk")
+				return true
+			}
+		}
+
+		if candidate.TestLocations(pk, locations) {
```

Comment:
Bloom filter may have better selectivity than other (range based?) filters, can we try put bloom filters first?